File: apps/web/src/api/matches.ts
import { z } from "zod";
import { invokeCommand, PaginationParams } from "@/lib/tauri-api";
import { createLogger } from "@/lib/logger";

const logger = createLogger("MatchesApi");

// ============================================================================
// Schemas
// ============================================================================

export const matchResponseSchema = z.object({
  id: z.string(),
  offer_id: z.string(),
  request_id: z.string(),
  score: z.string(), // Decimal as string
  confidence_band: z.string(),
  status: z.string(),
  operator_note: z.string().nullable(),
  medication_name: z.string(),
  offer_quantity: z.number(),
  request_quantity: z.number(),
  offer_price: z.string().nullable(), // Decimal as string
  max_price: z.string().nullable(), // Decimal as string
  created_at: z.coerce.date(),
  updated_at: z.coerce.date(),
});

export const matchListResponseSchema = z.object({
  matches: z.array(matchResponseSchema),
  total: z.number(),
});

export const matchStatsResponseSchema = z.object({
  total_matches: z.number(),
  pending_matches: z.number(),
  confirmed_matches: z.number(),
  rejected_matches: z.number(),
  auto_confirmed_matches: z.number(),
  avg_match_score: z.number(),
});

export type MatchResponse = z.infer<typeof matchResponseSchema>;
export type MatchListResponse = z.infer<typeof matchListResponseSchema>;
export type MatchStatsResponse = z.infer<typeof matchStatsResponseSchema>;

// ============================================================================
// Request Types
// ============================================================================

export type ListMatchesParams = {
  filter?: {
    status?: string;
  };
  pagination?: PaginationParams;
};

export type GetMatchParams = {
  id: string;
};

export type MatchActionParams = {
  id: string;
  data: {
    note?: string;
  };
};

// ============================================================================
// API Functions
// ============================================================================

/**
 * Get match statistics
 */
export async function getMatchStats(): Promise<MatchStatsResponse> {
  logger.info("Getting match statistics");
  return invokeCommand("get_match_stats", matchStatsResponseSchema);
}

/**
 * List matches with optional filtering and pagination
 */
export async function listMatches(
  params?: ListMatchesParams,
): Promise<MatchListResponse> {
  logger.info("Listing matches", params);
  return invokeCommand("list_matches", matchListResponseSchema, { params });
}

/**
 * Get a single match by ID
 */
export async function getMatch(id: string): Promise<MatchResponse> {
  logger.info("Getting match", { id });
  return invokeCommand("get_match", matchResponseSchema, {
    params: { id },
  });
}

/**
 * Confirm a match
 */
export async function confirmMatch(
  id: string,
  note?: string,
): Promise<MatchResponse> {
  logger.info("Confirming match", { id, note });
  return invokeCommand("confirm_match", matchResponseSchema, {
    params: {
      id,
      data: { note },
    },
  });
}

/**
 * Reject a match
 */
export async function rejectMatch(
  id: string,
  note?: string,
): Promise<MatchResponse> {
  logger.info("Rejecting match", { id, note });
  return invokeCommand("reject_match", matchResponseSchema, {
    params: {
      id,
      data: { note },
    },
  });
}

File: apps/web/src-tauri/src/ipc/commands/matches.rs
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Manager};
use utilities::{GetParams, ListParams};

use crate::ipc::response::IpcResponse;
use crate::state::AppState;
use db_service::types::MatchResponseDto;

// ============================================================================
// Filter Forms
// ============================================================================

#[derive(Debug, Deserialize)]
pub struct MatchFilter {
    status: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct MatchActionForm {
    note: Option<String>,
}

// ============================================================================
// Response Types
// ============================================================================

#[derive(Debug, Serialize)]
pub struct MatchListResponse {
    matches: Vec<MatchResponseDto>,
    total: usize,
}

#[derive(Debug, Serialize)]
pub struct MatchStatsResponse {
    pub total_matches: u64,
    pub pending_matches: u64,
    pub confirmed_matches: u64,
    pub rejected_matches: u64,
    pub auto_confirmed_matches: u64,
    pub avg_match_score: f64,
}

// ============================================================================
// Commands
// ============================================================================

/// Get match statistics
#[tauri::command]
pub async fn get_match_stats(app: AppHandle) -> IpcResponse<MatchStatsResponse> {
    async {
        let state = app.state::<AppState>();
        let stats_service = state.service_manager().stats_service();

        // Get dashboard stats which includes match statistics
        let dashboard_stats = stats_service.get_dashboard_stats().await?;

        // Calculate confirmed matches (total - pending - rejected - auto_confirmed)
        let confirmed_matches = dashboard_stats
            .total_matches
            .saturating_sub(dashboard_stats.pending_matches)
            .saturating_sub(dashboard_stats.auto_confirmed);

        Ok(MatchStatsResponse {
            total_matches: dashboard_stats.total_matches,
            pending_matches: dashboard_stats.pending_matches,
            confirmed_matches,
            rejected_matches: 0, // Not tracked separately in dashboard stats
            auto_confirmed_matches: dashboard_stats.auto_confirmed,
            avg_match_score: dashboard_stats.avg_match_score,
        })
    }
    .await
    .into()
}

/// List matches with optional filtering and pagination
#[tauri::command]
pub async fn list_matches(
    app: AppHandle,
    params: ListParams<MatchFilter>,
) -> IpcResponse<MatchListResponse> {
    async {
        let state = app.state::<AppState>();
        let matching_service = state.service_manager().matching_service();

        // Extract pagination parameters or use defaults
        let (page, page_size) = if let Some(pagination) = params.pagination() {
            (*pagination.page(), *pagination.page_size())
        } else {
            (0, 20) // Default: first page, 20 items
        };

        // Parse status filter if provided
        let status = if let Some(filter) = params.filter() {
            if let Some(status_str) = &filter.status {
                // Parse status string using FromStr trait
                let parsed_status = status_str
                    .parse::<db_service::entities::r#match::MatchStatus>()
                    .map_err(crate::error::AppError::ParseError)?;
                Some(parsed_status)
            } else {
                None
            }
        } else {
            None
        };

        // Use the matching service to list matches
        let matches = matching_service.list(status, page, page_size).await?;
        let total = matches.len();

        Ok(MatchListResponse { matches, total })
    }
    .await
    .into()
}

/// Get a single match by ID
#[tauri::command]
pub async fn get_match(app: AppHandle, params: GetParams) -> IpcResponse<MatchResponseDto> {
    async {
        let state = app.state::<AppState>();
        let matching_service = state.service_manager().matching_service();

        let match_result = matching_service.get_by_id(*params.id()).await?;

        Ok(match_result)
    }
    .await
    .into()
}

/// Confirm a match
#[tauri::command]
pub async fn confirm_match(
    app: AppHandle,
    params: utilities::UpdateParams<MatchActionForm>,
) -> IpcResponse<MatchResponseDto> {
    async {
        let state = app.state::<AppState>();
        let matching_service = state.service_manager().matching_service();

        use db_service::entities::r#match::MatchStatus;

        let form = params.data();
        let match_result = matching_service
            .update_status(*params.id(), MatchStatus::Confirmed, form.note.clone())
            .await?;

        Ok(match_result)
    }
    .await
    .into()
}

/// Reject a match
#[tauri::command]
pub async fn reject_match(
    app: AppHandle,
    params: utilities::UpdateParams<MatchActionForm>,
) -> IpcResponse<MatchResponseDto> {
    async {
        let state = app.state::<AppState>();
        let matching_service = state.service_manager().matching_service();

        use db_service::entities::r#match::MatchStatus;

        let form = params.data();
        let match_result = matching_service
            .update_status(*params.id(), MatchStatus::Rejected, form.note.clone())
            .await?;

        Ok(match_result)
    }
    .await
    .into()
}
